### DISCUSSION TITLE
What's our preferred UI development & test approach?

### DISCUSSION THREAD

**Discussion by jondavid-black** [DevOps-MBSE/AaC#116] [Q&A]
As we build up the UI MVP, we'll need good development and test approaches.  On the development side, I've been using GitPod which will not allow me to run "aac ui" successfully in a terminal.  We've been maintaining a GitPod dockerfile, and I really like working in GitPod, but it seems we'll need to do something different when working UI related issues.  Additionally, we'll need a good way to perform tests with the UI.  These tests need to be run in the GitHub Actions pipeline.  I don't have any experience testing Python UI applications and the top hits from a Google search didn't instill confidence that there's a broadly preferred solution.  

What recommendations do we want to make to contributors for development & testing going forward?  How do we want to document that recommendation?

**Comment by Coffee2Bits** [comment]
Gitpod is actually capable of developing the UI in, we just need to upgrade the image. I'm able to successfully run the UI in gitpod on my branch here: https://github.com/jondavid-black/AaC/pull/111 Gitpod has integrated a tool called NoVNC which exposes the VNC UI via a webserver, and I'm super impressed with it honestly. 

As far as testing, that's definitely a hard one, not only will we need a tool, but we'll need to make sure the CI workflow environment for the step is capable of running the UI and what test driver we can find -- I'm also finding few answers.

**Reply by Coffee2Bits** [reply]
Also going to add this Gitpod article for reference. It talks about Native UI development and even how to use X11 to run VSCode in Gitpod, but running VSCode in Gitpod is obviously an integrated feature now.

https://www.gitpod.io/blog/native-ui-with-vnc

**Comment by cvchaparro** [comment]
What are we specifically talking about with respect to testing the UI? Are we talking about testing button clicks, etc a-la Selenium or everything behind the UI, possibly ignoring testing the UI itself? Something else?

For actual UI testing, it seems like there's [a way](https://code.activestate.com/recipes/580751-testing-tkinter-or-selenium-for-tkinter/) to use Selenium (even though we're not developing this as a web app) with the addition of [rpyc](https://rpyc.readthedocs.io/en/latest/) but, admittedly, this seems like it could cause a bit of a maintenance headache, IMO.

Alternatively, the other tool I found was [pyautogui](https://pyautogui.readthedocs.io/en/latest/) that we might be able to use for this.
